File: pla/src/pla/EulerPhiAlgorithm.java
package pla;

import java.util.HashMap;
import java.util.Scanner;

public class EulerPhiAlgorithm {
	
	public static HashMap<Integer, Integer> get_factors(int num) {
		HashMap<Integer, Integer> factors = new HashMap<>();
		int numSqrt = (int) Math.sqrt(num);
		if (num % 2 == 1) {
			for (int i = 3; i <= numSqrt; i += 2) {
				while (num % i == 0 && Basic_Functions.isprime(i) && num != 1) {
					if (factors.containsKey(i)) {
						factors.replace(i, factors.get(i), factors.get(i) + 1);
					} else {
						factors.put(i, 1);
					}
					num /= i;
				}
			}
		} else {
			for (int i = 2; i <= numSqrt; i++) {
				while (num % i == 0 && Basic_Functions.isprime(i) && num != 1) {
					if (factors.containsKey(i)) {
						factors.replace(i, factors.get(i), factors.get(i) + 1);
					} else {
						factors.put(i, 1);
					}
					num /= i;
				}
			}
		}
		return factors;
	}
	
	
	public static int euler_phi(int num) {
		int phi = 1;
		if (Basic_Functions.isprime(num)) {
			phi = num - 1;
		} else {
			HashMap<Integer, Integer> factors = get_factors(num);
			for (int factor : factors.keySet()) {
				phi *= ((Math.pow(factor, factors.get(factor) - 1)) * (factor - 1));
			}
		}
		return phi;
	}

	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		System.out.print("Number: ");
		int num = sc.nextInt();
		sc.close();
		
		System.out.println("Phi value: " + euler_phi(num));
	}

}

File: python/EulerPhiAlgorithm.py
# To find Phi value of some no.
from Basic_Functions import isprime, get_integer
from math import sqrt

def get_factors(num):
	factors = dict()
	if num % 2:			# odd
		for i in range(3, int(sqrt(num)) + 1, 2):
			while num % i == 0 and isprime(i) and num != 1:
				factors[i] = factors.get(i, 0) + 1
				num /= i
	else:
		for i in range(2, int(sqrt(num)) + 1):
			while num % i == 0 and isprime(i) and num != 1:
				factors[i] = factors.get(i, 0) + 1
				num /= i
	return factors

def euler_phi(num):
	phiVal = 1
	if isprime(num):
		phiVal = num - 1
	else:
		factors = get_factors(num)
		for factor in factors:
			phiVal *= ((factor ** (factors[factor] - 1)) * (factor - 1))
	return phiVal


no = get_integer('Number: ')
print('Phi value: ' + str(euler_phi(no)))
